libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,937 @@
+
+/// Helpers for a given storage configuration.
+#[derive(Clone, Debug)]
+pub struct Format {
+    /// The size in bytes of a page in the storage.
+    ///
+    /// # Invariant
+    ///
+    /// - Words divide a page evenly.
+    /// - There are at least 8 words in a page.
+    /// - There are at most `MAX_PAGE_SIZE` bytes in a page.
+    page_size: usize,
+
+    /// The number of pages in the storage.
+    ///
+    /// # Invariant
+    ///
+    /// - There are at least 3 pages.
+    /// - There are at most `MAX_PAGE_INDEX + 1` pages.
+    num_pages: usize,
+
+    /// The maximum number of times a page can be erased.
+    ///
+    /// # Invariant
+    ///
+    /// - A page can be erased at most `MAX_ERASE_CYCLE` times.
+    max_page_erases: usize,
+}
+
+impl Format {
+    /// Extracts the format from a storage.
+    ///
+    /// Returns `None` if the storage is not [supported].
+    ///
+    /// [supported]: struct.Format.html#method.is_storage_supported
+    pub fn new<S: Storage>(storage: &S) -> Option<Format> {
+        if Format::is_storage_supported(storage) {
+            Some(Format {
+                page_size: storage.page_size(),
+                num_pages: storage.num_pages(),
+                max_page_erases: storage.max_page_erases(),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns whether a storage is supported.
+    ///
+    /// A storage is supported if the following conditions hold:
+    /// - The size of a word is 4 bytes.
+    /// - The size of a word evenly divides the size of a page.
+    /// - A page contains at least 8 words.
+    /// - A page contains at most [`MAX_PAGE_SIZE`] bytes.
+    /// - There are at least 3 pages.
+    /// - There are at most [`MAX_PAGE_INDEX`]` + 1` pages.
+    /// - A word can be written at least twice between erase cycles.
+    /// - The maximum number of erase cycles is at most [`MAX_ERASE_CYCLE`].
+    ///
+    /// [`MAX_PAGE_SIZE`]: constant.MAX_PAGE_SIZE.html
+    /// [`MAX_PAGE_INDEX`]: constant.MAX_PAGE_INDEX.html
+    /// [`MAX_ERASE_CYCLE`]: constant.MAX_ERASE_CYCLE.html
+    fn is_storage_supported<S: Storage>(storage: &S) -> bool {
+        let word_size = storage.word_size();
+        let page_size = storage.page_size();
+        let num_pages = storage.num_pages();
+        let max_word_writes = storage.max_word_writes();
+        let max_page_erases = storage.max_page_erases();
+        word_size == 4
+            && page_size % word_size == 0
+            && (8 * word_size <= page_size && page_size <= MAX_PAGE_SIZE)
+            && (3 <= num_pages && num_pages <= MAX_PAGE_INDEX + 1)
+            && max_word_writes >= 2
+            && max_page_erases <= MAX_ERASE_CYCLE
+    }
+
+    /// The size of a word in bytes.
+    pub fn word_size(&self) -> usize {
+        WORD_SIZE
+    }
+
+    /// The size of a page in bytes.
+    ///
+    /// We have `32 <= self.page_size() <= MAX_PAGE_SIZE` assuming a word is 4 bytes.
```

Comment:
Here, wouldn't "We have `8 * self.word_size() <= self.page_size() <= MAX_PAGE_SIZE`" work?

Reply:
Yes it's the same. I thought inlining the constants might help the reader get an idea of actual values, but if it's more confusing let's use the exact formulas. Done through the whole file introducing more constants.